internal/webhook/rule/image_digest.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 The casbin Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package rule
+
+import (
+	"encoding/json"
+	"fmt"
+	v1 "k8s.io/api/admission/v1"
+	app "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	"log"
+	"regexp"
+)
+
+/*
+[ks-admission-general-image-digest]
+
+<Brief introduction>
+Only images with a valid digest can be allowed.
+
+<Coverage Area of Rule>
+pods and deployments resources will be checked.
+*/
+func (g *Rules) ImageDigest(review *v1.AdmissionReview, model string, policy string) error {
+	var resourceKind = review.Request.Resource.Resource
+	//fmt.Println(resourceKind)
+	switch resourceKind {
+	case "pods":
+		return g.imageDigestForPod(review, model, policy)
+	case "deployments":
+		return g.imageDigestForDeployment(review, model, policy)
+	default:
+		return nil
+	}
+}
+
+func (g *Rules) imageDigestForPod(review *v1.AdmissionReview, model string, policy string) error {
+
+	if review.Request.Operation == "DELETE" {
+		//delete operation have no docker image to check
+		log.Printf("ImageDigest: pod %s:%s approved", review.Request.Namespace, review.Request.Name)
+		return nil
+	}
+	var podObject core.Pod
+	if err := json.Unmarshal(review.Request.Object.Raw, &podObject); err != nil {
+		log.Printf("ImageDigest: pod %s:%s rejected due to error:%s", review.Request.Namespace, review.Request.Name, err.Error())
+		return err
+	}
+	allContainers := make([]core.Container, len(podObject.Spec.Containers))
+	copy(allContainers, podObject.Spec.Containers)
+	allContainers = append(allContainers, podObject.Spec.InitContainers...)
+	reg, err := regexp.Compile(`@[a-z0-9]+([+._-][a-z0-9]+)*:[a-zA-Z0-9=_-]+`)
```

Comment:
`regexp.Compile(`@[a-z0-9]+([+._-][a-z0-9]+)*:[a-zA-Z0-9=_-]+`)` Yes As a field of the package, it does not need to be compiled every called.